[scroll-animations] allow timeline ranges to be used when specifying keyframes

graouts opened the pull request:
#### 21ebfdcf906c11c01ead1af39be3fa7de873002e
<pre>
[scroll-animations] allow timeline ranges to be used when specifying keyframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=286920">https://bugs.webkit.org/show_bug.cgi?id=286920</a>

Reviewed by Sam Weinig.

The Scroll-driven Animations specification extends the definition of keyframes
to allow for timeline ranges to be used on top of the existing percentage values.
This means that for view timelines, authors can specify specific points within
that timeline&apos;s range as keyframe offsets. For instance, here is an animation
which would animate the opacity from 0 to 1 as the target element enters the
visible range within its scrolling container, and then from 1 to 0 as it leaves
it:

```
@keyframes my-anim {
    cover 0% { opacity: 0 }
    contain 0% { opacity: 1 }
    contain 100% { opacity: 1 }
    cover 100% { opacity: 0 }
}
```

This is a pretty big change since until now, keyframe offsets were either implicit
omitted values or explicit numeric values that could synchronously be computed to
a finite value. With timeline ranges, computed offsets will be NaN until we have
viable metrics for the view timeline&apos;s source and subject and will dynamically
update as the view timeline&apos;s metrics and scroll offset changes.

To that end, we introduce the notion of a `specifiedOffset` on `BlendingKeyframe`
and make the existing `offset` be the computed value. Specified offsets are
represented via the new `BlendingKeyframe::Offset` struct which encompasses the
range name (including `Omitted` and `Normal` values) and the numeric value.
This provides a way to determine whether a specified value was implicit or
explicit.

As for computed values, they are all NaN by default and resolved using the new
`BlendingKeyframes::updatedComputedOffsets()` method which will let the caller
provide a callback to convert each specified offset, leaving the mapping from
the view timeline range to keyframe effects.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-view-functional-notation.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/merge-timeline-offset-keyframes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-in-keyframe-change-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-hidden-subject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-hidden-subject.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-with-document-timeline-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-range-name-offset-in-keyframes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-keyframe-boundary-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-keyframe-boundary-interpolation.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/contain-alignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/get-keyframes-with-timeline-offset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/timeline-offset-in-keyframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-source.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/zero-intrinsic-iteration-duration.tentative-expected.txt:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::animationProgressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::insert):
(WebCore::BlendingKeyframes::copyKeyframes):
(WebCore::zeroPercentKeyframe):
(WebCore::hundredPercentKeyframe):
(WebCore::BlendingKeyframes::fillImplicitKeyframes):
(WebCore::BlendingKeyframes::analyzeKeyframe):
(WebCore::BlendingKeyframes::updatedComputedOffsets):
(WebCore::BlendingKeyframe::BlendingKeyframe):
(WebCore::BlendingKeyframe::usesRangeOffset const):
* Source/WebCore/animation/BlendingKeyframes.h:
(WebCore::BlendingKeyframes::hasKeyframeNotUsingRangeOffset const):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::rangeStringToSingleTimelineRangeName):
(WebCore::isTimelineRangeOffsetValid):
(WebCore::rangeStringFromSingleTimelineRangeName):
(WebCore::doubleOrTimelineRangeOffsetFromString):
(WebCore::validateKeyframeOffset):
(WebCore::computedOffset):
(WebCore::computeMissingKeyframeOffsets):
(WebCore::processKeyframeLikeObject):
(WebCore::processIterableKeyframes):
(WebCore::processPropertyIndexedKeyframes):
(WebCore::timelineRangeOffsetFromSpecifiedOffset):
(WebCore::KeyframeEffect::getKeyframes):
(WebCore::KeyframeEffect::processKeyframes):
(WebCore::specifiedOffsetForParsedKeyframe):
(WebCore::KeyframeEffect::updateBlendingKeyframes):
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
(WebCore::KeyframeEffect::animationTimelineDidChange):
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):
(WebCore::KeyframeEffect::animationDidTick):
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
(WebCore::KeyframeEffect::computeHasImplicitKeyframeForAcceleratedProperty):
(WebCore::KeyframeEffect::activeViewTimeline):
(WebCore::KeyframeEffect::animationProgressBasedTimelineSourceDidChangeMetrics):
(WebCore::KeyframeEffect::updateComputedKeyframeOffsetsIfNeeded):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffect.idl:
* Source/WebCore/animation/KeyframeInterpolation.cpp:
(WebCore::KeyframeInterpolation::interpolationKeyframes const):
* Source/WebCore/animation/KeyframeInterpolation.h:
(WebCore::KeyframeInterpolation::Keyframe::hasResolvedOffset const):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::create):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::create):
(WebCore::ViewTimeline::intervalForTimelineRangeName const):
(WebCore::ViewTimeline::mapOffsetToTimelineRange const):
(WebCore::ViewTimeline::offsetIntervalForTimelineRangeName const):
(WebCore::ViewTimeline::offsetIntervalForAttachmentRange const):
(WebCore::ViewTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::progressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::Key::writeToString const):
(WebCore::StyleRuleKeyframe::StyleRuleKeyframe):
(WebCore::StyleRuleKeyframe::create):
(WebCore::StyleRuleKeyframe::keyText const):
(WebCore::StyleRuleKeyframe::setKeyText):
* Source/WebCore/css/CSSKeyframeRule.h:
* Source/WebCore/css/CSSKeyframesRule.cpp:
(WebCore::StyleRuleKeyframes::findKeyframeIndex const):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseKeyframeKeyList):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseKeyframeKeyList):
(WebCore::CSSParserImpl::consumeKeyframeStyleRule):
(WebCore::CSSParserImpl::consumeKeyframeKeyList): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
(WebCore::CSSPropertyParserHelpers::consumeKeyframeKeyList):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::isAnimationRangeKeyword):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/style/StyleResolver.cpp:
(WTF::StyleRuleKeyframeKeyHash::hash):
(WTF::StyleRuleKeyframeKeyHash::equal):
(WTF::HashTraits&lt;WebCore::StyleRuleKeyframe::Key&gt;::emptyValue):
(WTF::HashTraits&lt;WebCore::StyleRuleKeyframe::Key&gt;::isEmptyValue):
(WTF::HashTraits&lt;WebCore::StyleRuleKeyframe::Key&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebCore::StyleRuleKeyframe::Key&gt;::isDeletedValue):
(WebCore::Style::Resolver::keyframeRulesForName const):
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebCore/style/StyleResolver.h:

Canonical link: <a href="https://commits.webkit.org/289948@main">https://commits.webkit.org/289948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55bca165891c7b679b03a896c0d6c7b5d23be65a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88471 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93429 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39225 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90522 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8377 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16174 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68239 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25954 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6413 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80024 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48605 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6185 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34438 "Found 1 new test failure: editing/undo/redo-reapply-edit-command-crash.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38333 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76558 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35323 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95271 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15646 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11467 "Found 3 new test failures: http/wpt/mediarecorder/MediaRecorder-matroska-AV-audio-video-dataavailable.html imported/w3c/web-platform-tests/screen-orientation/lock-basic.html pdf/annotations/radio-buttons-select-second.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77104 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15902 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75880 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76365 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20752 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19172 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8685 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13836 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15662 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18852 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17185 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.h:
```diff
@@ -35,6 +35,8 @@ struct CSSParserContext;
 
 namespace CSSPropertyParserHelpers {
 
+Vector<std::pair<CSSValueID, double>> consumeKeyframeKeyList(CSSParserTokenRange&);
```
Assuming this is meant to be parsing the updated production of `<keyframe-selector>` defined https://drafts.csswg.org/scroll-animations-1/#named-range-keyframes, would you mind adding production name and URL as done below?

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
```diff
@@ -37,6 +37,48 @@
+            if (id == CSSValueNormal)
+                return { };
+            auto& token = range.consumeIncludingWhitespace();
+            if (token.type() == PercentageToken && token.numericValue() >= 0 && token.numericValue() <= 100)
```
Assuming this is for https://drafts.csswg.org/scroll-animations-1/#named-range-keyframes, these are supposed to be `<percentage [0,100]>` which include calc(), so should either use `consumePercentage()` or `MetaConsumer<CSS::Percentage<CSS::Range{0,100}>>`.

If the spec intends for calc() not to be included, it can use the production `<percentage-token>` (https://drafts.csswg.org/css-values-5/#typedef-progress is an example of `<percentage-token>` being used).

graouts replied:
Thanks for the tip! Using `consumePercentage()` and `CSSPrimitiveValue::resolveAsPercentage()` made an additional WPT test pass which tested a `calc(20% + 30%)` value which previously caused a parsing error.

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
```diff
@@ -37,6 +37,67 @@
+    // to return a CSS value that can be kept as-is on a BlendingKeyframe so that
+    // resolution happens when we have the necessary context when the keyframes are
+    // associated with a target element.
+    CSSToLengthConversionData conversionData { };
```
Instead of using an empty one of these, please use `percentageValue->resolveAsPercentageDeprecated()` as that is easier to search for.

weinig commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -63,9 +63,14 @@ class ViewTimeline final : public ScrollTimeline {
 private:
     ScrollTimeline::Data computeTimelineData() const final;
-    std::pair<WebAnimationTime, WebAnimationTime> intervalForAttachmentRange(const TimelineRange&) const final;
+    double mapOffsetToTimelineRange(const ScrollTimeline::Data&, const SingleTimelineRange::Name, const Function<double(const float&)>&) const;
```
There is should probably be a template function:

```
template<typename F> double mapOffsetToTimelineRange(const ScrollTimeline::Data&, const SingleTimelineRange::Name, F&&) const;
```

All the callers are in the same file, so there is no reason to use WTF::Function, which can allocate.

weinig commented on Source/WebCore/animation/KeyframeInterpolation.cpp:
```diff
@@ -46,7 +46,7 @@ const KeyframeInterpolation::KeyframeInterval KeyframeInterpolation::interpolati
         auto& keyframe = keyframeAtIndex(i);
         auto offset = keyframe.offset();
-        if (!keyframe.animatesProperty(property))
+        if (std::isnan(offset) || !keyframe.animatesProperty(property))
```
This use of std::isnan() seems so surprising here. I think a function (`bool isPlaceholderOffset` or whatever) would go a long way to make this more self documenting.

weinig commented on Source/WebCore/animation/BlendingKeyframes.cpp:
```diff
@@ -379,10 +384,45 @@ void BlendingKeyframes::analyzeKeyframe(const BlendingKeyframe& keyframe)
+    for (auto& keyframe : m_keyframes)
+        keyframe.setComputedOffset(callback(keyframe.specifiedOffset()));
+
+    std::stable_sort(m_keyframes.begin(), m_keyframes.end(), [](auto& lhs, auto& rhs) {
```
You should be able to use `std::ranges::stable_sort(m_keyframes, ...)` here.

graouts committed:
[scroll-animations] allow timeline ranges to be used when specifying keyframes
https://bugs.webkit.org/show_bug.cgi?id=286920

Reviewed by Sam Weinig.

The Scroll-driven Animations specification extends the definition of keyframes
to allow for timeline ranges to be used on top of the existing percentage values.
This means that for view timelines, authors can specify specific points within
that timeline's range as keyframe offsets. For instance, here is an animation
which would animate the opacity from 0 to 1 as the target element enters the
visible range within its scrolling container, and then from 1 to 0 as it leaves
it:

```
@keyframes my-anim {
    cover 0% { opacity: 0 }
    contain 0% { opacity: 1 }
    contain 100% { opacity: 1 }
    cover 100% { opacity: 0 }
}
```

This is a pretty big change since until now, keyframe offsets were either implicit
omitted values or explicit numeric values that could synchronously be computed to
a finite value. With timeline ranges, computed offsets will be NaN until we have
viable metrics for the view timeline's source and subject and will dynamically
update as the view timeline's metrics and scroll offset changes.

To that end, we introduce the notion of a `specifiedOffset` on `BlendingKeyframe`
and make the existing `offset` be the computed value. Specified offsets are
represented via the new `BlendingKeyframe::Offset` struct which encompasses the
range name (including `Omitted` and `Normal` values) and the numeric value.
This provides a way to determine whether a specified value was implicit or
explicit.

As for computed values, they are all NaN by default and resolved using the new
`BlendingKeyframes::updatedComputedOffsets()` method which will let the caller
provide a callback to convert each specified offset, leaving the mapping from
the view timeline range to keyframe effects.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/animation-timeline-view-functional-notation.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/merge-timeline-offset-keyframes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-in-keyframe-change-timeline.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-hidden-subject-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-hidden-subject.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-offset-keyframes-with-document-timeline-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/timeline-range-name-offset-in-keyframes.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-keyframe-boundary-interpolation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-keyframe-boundary-interpolation.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/contain-alignment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/get-keyframes-with-timeline-offset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/timeline-offset-in-keyframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/view-timeline-source.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/zero-intrinsic-iteration-duration.tentative-expected.txt:
* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::animationProgressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/BlendingKeyframes.cpp:
(WebCore::BlendingKeyframes::insert):
(WebCore::BlendingKeyframes::copyKeyframes):
(WebCore::zeroPercentKeyframe):
(WebCore::hundredPercentKeyframe):
(WebCore::BlendingKeyframes::fillImplicitKeyframes):
(WebCore::BlendingKeyframes::analyzeKeyframe):
(WebCore::BlendingKeyframes::updatedComputedOffsets):
(WebCore::BlendingKeyframe::BlendingKeyframe):
(WebCore::BlendingKeyframe::usesRangeOffset const):
* Source/WebCore/animation/BlendingKeyframes.h:
(WebCore::BlendingKeyframes::hasKeyframeNotUsingRangeOffset const):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::rangeStringToSingleTimelineRangeName):
(WebCore::isTimelineRangeOffsetValid):
(WebCore::rangeStringFromSingleTimelineRangeName):
(WebCore::doubleOrTimelineRangeOffsetFromString):
(WebCore::validateKeyframeOffset):
(WebCore::computedOffset):
(WebCore::computeMissingKeyframeOffsets):
(WebCore::processKeyframeLikeObject):
(WebCore::processIterableKeyframes):
(WebCore::processPropertyIndexedKeyframes):
(WebCore::timelineRangeOffsetFromSpecifiedOffset):
(WebCore::KeyframeEffect::getKeyframes):
(WebCore::KeyframeEffect::processKeyframes):
(WebCore::specifiedOffsetForParsedKeyframe):
(WebCore::KeyframeEffect::updateBlendingKeyframes):
(WebCore::KeyframeEffect::setBlendingKeyframes):
(WebCore::KeyframeEffect::computeCSSAnimationBlendingKeyframes):
(WebCore::KeyframeEffect::animationTimelineDidChange):
(WebCore::KeyframeEffect::setAnimatedPropertiesInStyle):
(WebCore::KeyframeEffect::animationDidTick):
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
(WebCore::KeyframeEffect::computeHasImplicitKeyframeForAcceleratedProperty):
(WebCore::KeyframeEffect::activeViewTimeline):
(WebCore::KeyframeEffect::animationProgressBasedTimelineSourceDidChangeMetrics):
(WebCore::KeyframeEffect::updateComputedKeyframeOffsetsIfNeeded):
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/KeyframeEffect.idl:
* Source/WebCore/animation/KeyframeInterpolation.cpp:
(WebCore::KeyframeInterpolation::interpolationKeyframes const):
* Source/WebCore/animation/KeyframeInterpolation.h:
(WebCore::KeyframeInterpolation::Keyframe::hasResolvedOffset const):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::create):
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::create):
(WebCore::ViewTimeline::intervalForTimelineRangeName const):
(WebCore::ViewTimeline::mapOffsetToTimelineRange const):
(WebCore::ViewTimeline::offsetIntervalForTimelineRangeName const):
(WebCore::ViewTimeline::offsetIntervalForAttachmentRange const):
(WebCore::ViewTimeline::intervalForAttachmentRange const):
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::progressBasedTimelineSourceDidChangeMetrics):
* Source/WebCore/css/CSSKeyframeRule.cpp:
(WebCore::StyleRuleKeyframe::Key::writeToString const):
(WebCore::StyleRuleKeyframe::StyleRuleKeyframe):
(WebCore::StyleRuleKeyframe::create):
(WebCore::StyleRuleKeyframe::keyText const):
(WebCore::StyleRuleKeyframe::setKeyText):
* Source/WebCore/css/CSSKeyframeRule.h:
* Source/WebCore/css/CSSKeyframesRule.cpp:
(WebCore::StyleRuleKeyframes::findKeyframeIndex const):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseKeyframeKeyList):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseKeyframeKeyList):
(WebCore::CSSParserImpl::consumeKeyframeStyleRule):
(WebCore::CSSParserImpl::consumeKeyframeKeyList): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
(WebCore::CSSPropertyParserHelpers::consumeKeyframeKeyList):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::isAnimationRangeKeyword):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/style/StyleResolver.cpp:
(WTF::StyleRuleKeyframeKeyHash::hash):
(WTF::StyleRuleKeyframeKeyHash::equal):
(WTF::HashTraits<WebCore::StyleRuleKeyframe::Key>::emptyValue):
(WTF::HashTraits<WebCore::StyleRuleKeyframe::Key>::isEmptyValue):
(WTF::HashTraits<WebCore::StyleRuleKeyframe::Key>::constructDeletedValue):
(WTF::HashTraits<WebCore::StyleRuleKeyframe::Key>::isDeletedValue):
(WebCore::Style::Resolver::keyframeRulesForName const):
(WebCore::Style::Resolver::keyframeStylesForAnimation):
* Source/WebCore/style/StyleResolver.h:

Canonical link: https://commits.webkit.org/289948@main
